### PR TITLE
Swing action invoke and wait

### PR DESCRIPTION
**Pull request by DanVanAtta**
- Cleanup for TripleA.java
- Add new library method SwingAction.invokeAndWait, simplifies and replaces SwingUtilities.invokeAndWait
- Allow SwingUtilities to check the current thread. We had quite a bit of this pattern:
```
Runnable r = 
if( Swing.eventThread) {
 r.run();
} else {
  SwingUtilities.invokeAndWait(r)
}
```

The cost for a new thread is very cheap these days. So this can be re-written as:
```
Runnable r = ...
SwingUtilities.invokeAndWait(r)
```
Actually, not only is the cost of a new thread expensive, but the current thread check is pretty much what SwingUtilities does, so  wrapping our own is redundant.

[Review on Reviewable](https://reviewable.io/reviews/triplea-game/triplea/514)

### CODE REVIEW

**Comment by DanVanAtta** [comment]
merge conflicts fixed, ready for review again

File: src/games/strategy/engine/framework/GameRunner2.java [hunk 1]
```diff
@@ -149,6 +150,19 @@ public static void main(final String[] args) {
       ErrorHandler.registerExceptionHandler();
     })).start();
 
+    s_countDownLatch = new CountDownLatch(1);
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner2.java]: Bad merge resolution here, I need to fix this and remove it.

File: src/games/strategy/engine/framework/ui/NewGameChooserModel.java [hunk 2]
```diff
@@ -205,8 +192,8 @@ private static void addNewGameChooserEntry(final List<NewGameChooserEntry> entri
     } catch (final EngineVersionException e) {
       System.out.println(e.getMessage());
     } catch (final SAXParseException e) {
-      System.err
-          .println("Could not parse:" + uri + " error at line:" + e.getLineNumber() + " column:" + e.getColumnNumber());
+      String msg = "Could not parse:" + uri + " error at line:" + e.getLineNumber() + " column:" + e.getColumnNumber();
```
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/ui/NewGameChooserModel.java]: opportunistic formatting update here.

File: src/games/strategy/engine/random/PBEMDiceRoller.java [hunk 3]
```diff
@@ -82,7 +73,7 @@ public void run() {
     return dialog.getDiceRoll();
   }
 
-  private Frame getFocusedFrame() {
+  private static Frame getFocusedFrame() {
```
Comment (DanVanAtta) [inline src/games/strategy/engine/random/PBEMDiceRoller.java]: opportunistic compiler warning fix

File: src/games/strategy/triplea/TripleA.java [hunk 4]
```diff
@@ -43,15 +40,17 @@
...
   public static final String HUMAN_PLAYER_TYPE = "Human";
   public static final String WEAK_COMPUTER_PLAYER_TYPE = "Easy (AI)";
   public static final String FAST_COMPUTER_PLAYER_TYPE = "Fast (AI)";
   public static final String PRO_COMPUTER_PLAYER_TYPE = "Hard (AI)";
   public static final String DOESNOTHINGAI_COMPUTER_PLAYER_TYPE = "Does Nothing (AI)";
-  // public static final String NONE = "None (AI)";
-  protected transient ITripleaDisplay m_display;
+  protected transient ITripleaDisplay display;
+
+  public TripleA() {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/TripleA.java]: opportunistic improvement. For a longer class, at least one explicit constructor is useful to make the constructors easy to find (rather than not finding them and deducing that there are none)

**Comment by DanVanAtta** [comment]
Fixed bad merge resolutions. Ready for merge if things look good in review.

**Comment by ron-murhammer** [comment]
@DanVanAtta Appears to be some issues with this. Did you test it after the changes? I get this error when I launch the game from GameRunner2.java:

```
Exception in thread "AWT-EventQueue-0" java.lang.Error: Cannot call invokeAndWait from the event dispatcher thread
	at java.awt.EventQueue.invokeAndWait(EventQueue.java:1301)
	at java.awt.EventQueue.invokeAndWait(EventQueue.java:1294)
	at javax.swing.SwingUtilities.invokeAndWait(SwingUtilities.java:1348)
	at games.strategy.common.swing.SwingAction.invokeAndWait(SwingAction.java:59)
	at games.strategy.engine.framework.GameRunner2.setupLookAndFeel(GameRunner2.java:245)
	at games.strategy.engine.framework.GameRunner2.lambda$0(GameRunner2.java:143)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:756)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:726)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82) 
```

It also doesn't appear to apply the appropriate skin as the menu is a light gray instead of the usual dark gray:

capture

**Comment by DanVanAtta** [comment]
@ron-murhammer , I recall seeing that, I think it's a result of bad merge conflict resolution and a change was lost.

**Comment by DanVanAtta** [comment]
Waiting for the travis tests to re-verify, otherwise I just pushed a fix for the latest problem. Was a merge conflict that was not resolved correctly, so was pretty easy to get back into working state.

File: src/games/strategy/engine/framework/GameRunner2.java [hunk 5]
```diff
@@ -241,23 +242,24 @@ private static String getValue(final String arg) {
...
-      // Then it is because you included the swingx substance library without including swingx.
-      // You can solve by including both swingx libraries or removing both,
-      // or by setting the look and feel twice in a row.
-    } catch (final Throwable t) {
-      System.out.println("LOOK AND FEEL exception: " + t.getMessage());
-      if (!GameRunner.isMac()) {
-        try {
-          UIManager.setLookAndFeel(UIManager.getSystemLookAndFeelClassName());
-        } catch (final Exception e) {
+    SwingAction.invokeAndWait(() -> {
```
Comment (ron-murhammer) [inline src/games/strategy/engine/framework/GameRunner2.java]: Can this block be cleaned up some to make it more readable? Lots of nesting makes this hard to follow.

File: src/games/strategy/engine/framework/ui/NewGameChooserModel.java [hunk 6]
```diff
@@ -149,45 +152,33 @@ private static ZipProcessingResult processZipEntry(final URLClassLoader loader,
    * Open up a confirmation dialog, if user says yes, delete the map specified by
    * parameter, then show confirmation of deletion.
    */
-  private static void confirmWithUserAndThenDeleteCorruptZipFile(final File map, Optional<String> details) {
```
Comment (ron-murhammer) [inline src/games/strategy/engine/framework/ui/NewGameChooserModel.java]: What is this change for? Hard to follow as the whole method is pretty much removed/added in the diff. Not sure if this is a bad merge or not?

File: src/games/strategy/triplea/TripleA.java [hunk 7]
```diff
@@ -79,93 +78,69 @@
...
+      game.removeDisplay(display);
+      display.shutDown();
+      display = null;
     }
   }
 
-  protected void initializeGame() {}
 
   @Override
-  public void startGame(final IGame game, final Set<IGamePlayer> players, final boolean headless) throws Exception {
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/TripleA.java]: This diff is a little hard to follow. What are the main changes here?

File: src/games/strategy/triplea/ui/BattlePanel.java [hunk 8]
```diff
@@ -241,61 +242,52 @@ public void run() {
     }
   }
 
-  public void showBattle(final GUID battleID, final Territory location, final String battleTitle,
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/BattlePanel.java]: Why was the method's parameters changed here? Were they unused?

File: src/games/strategy/triplea/ui/TripleAFrame.java [hunk 9]
```diff
@@ -1429,22 +1409,7 @@ public PoliticalActionAttachment getPoliticalActionChoice(final PlayerID player,
...
-          }
-        });
-      } catch (final Exception e) {
-        e.printStackTrace();
-      }
-    } else {
-      requestFocusInWindow();
-      transferFocus();
-    }
+    requestWindowFocus();
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/TripleAFrame.java]: Do we not need the invokeAndWait here? Also looks like transferFocus() was removed?

File: src/games/strategy/triplea/ui/TripleAFrame.java [hunk 10]
```diff
@@ -1455,22 +1420,7 @@ public UserActionAttachment getUserActionChoice(final PlayerID player, final boo
...
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/TripleAFrame.java]: Do we not need the invokeAndWait here? Also looks like transferFocus() was removed?

File: src/games/strategy/triplea/ui/TripleAFrame.java [hunk 11]
```diff
@@ -1481,22 +1431,7 @@ public TechRoll getTechRolls(final PlayerID id) {
...
```
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/TripleAFrame.java]: Do we not need the invokeAndWait here? Also looks like transferFocus() was removed?

**Comment by ron-murhammer** [comment]
@DanVanAtta - Added some comments/questions. Tested functionality and seems to launch and show everything properly now.

File: src/games/strategy/engine/framework/ui/NewGameChooserModel.java (hunk 6 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/ui/NewGameChooserModel.java]: The change is I think as intended. Not sure why the method signature line shows an update, that does not look changed. Otherwise we have a method that was previously wrapped entirely in a try/catch unnecessarily. Removing that, then worked in the library functions to remove the rest of the try/catches.

File: src/games/strategy/triplea/TripleA.java (hunk 7 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/TripleA.java]: The big try catch around the method is removed. Then the ad-hoc error handling is all removed as well since the library function now handles the error handling in a unified way. It was interesting to see the variations of how we handled Interrupted exception, sometime it was fully handled very formally, other time it was just ignored, and ranges in between.

File: src/games/strategy/triplea/ui/BattlePanel.java (hunk 8 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/BattlePanel.java]: Yes, unused.

File: src/games/strategy/engine/framework/GameRunner2.java (hunk 5 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner2.java]: The catch block is when the nesting starts to get bad. The one extra level practically does not matter here. The main goal of this PR is the SwingAction update, I'm leaving this specific clean up out of scope.

File: src/games/strategy/triplea/ui/TripleAFrame.java (hunk 9 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/TripleAFrame.java]: The method 'requestWindowFocus()' handles it al:
```
  private void requestWindowFocus() {
    SwingAction.invokeAndWait(() -> {
      requestFocusInWindow();
      transferFocus();
    });
  }
```

File: src/games/strategy/engine/framework/GameRunner2.java (hunk 5 above)
Comment (DanVanAtta) [inline src/games/strategy/engine/framework/GameRunner2.java]: I took another look at this, looking at the source code, the indentation is actually not that bad.
The biggest problem is the control flow via exception handling, then the completely ignored exception at the end.

Simply moving stuff around to reduce the indentation is not a good fix, since there is very little here in this method. What more needs to be done is to enforce that the substance lib is actually included, and to remove the Mac special casing, and last, add a verification check to see which default look and feel shoudl be used, rather than blindly trying one, failing silently, then trying another and silently failing again.

File: src/games/strategy/triplea/ui/TripleAFrame.java (hunk 9 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ui/TripleAFrame.java]: Oh totally missed it was a new method. Just saw it started with 'request' and thought it was the same one.

**Comment by ron-murhammer** [comment]
@DanVanAtta Ok. Merging this now. I will say most of your more recent PRs are much cleaner and more focused than this one. This one kind of has a few things outside of the main SwingAction.invokeAndWait refactor that made it a little more difficult to review as it was jumping around some. Definitely prefer the more focused/smaller PRs when possible.
